File: tools/generate_test_cases.py

import requests
import json
import os
import sys
import argparse
from datetime import datetime
from dotenv import load_dotenv

# Load env
load_dotenv()

OLLAMA_URL = os.getenv("OLLAMA_URL", "http://localhost:11434/api/chat")
MODEL = os.getenv("MODEL", "llama3.2")

SYSTEM_PROMPT = """You are an expert QA Automation Engineer.
Your task is to generate detailed test cases based on the user's requirement.
Output MUST be a strictly valid JSON object with the following structure:
{
  "testCases": [
    {
      "id": "TC_{number}",
      "title": "{action} should {expected_result} when {condition}",
      "preconditions": ["List of preconditions"],
      "steps": ["Step 1", "Step 2"],
      "expected_result": "Outcome",
      "priority": "High|Medium|Low",
      "type": "Positive|Negative"
    }
  ]
}
Ensure the JSON is valid. Do not include any explanations or text outside the JSON object."""

def generate_test_cases(prompt):
    messages = [
        {"role": "system", "content": SYSTEM_PROMPT},
        {"role": "user", "content": prompt}
    ]
    
    payload = {
        "model": MODEL,
        "messages": messages,
        "stream": False,
        "format": "json"
    }
    
    try:
        response = requests.post(OLLAMA_URL, json=payload)
        response.raise_for_status()
        data = response.json()
        content = data.get("message", {}).get("content", "{}")
        
        # Parse to verify JSON
        json_content = json.loads(content)
        
        # Save to .tmp (Self-Healing/Audit)
        tmp_dir = os.path.join(os.getcwd(), '.tmp')
        os.makedirs(tmp_dir, exist_ok=True)
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        filepath = os.path.join(tmp_dir, f"test_cases_{timestamp}.json")
        
        with open(filepath, 'w') as f:
            json.dump(json_content, f, indent=2)
            
        print(json.dumps(json_content, indent=2))
        return filepath
        
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("prompt", help="User prompt for test case generation")
    args = parser.parse_args()
    generate_test_cases(args.prompt)

File: tools/generate_test_cases.js

import { fetch } from 'undici'; // Use global fetch if Node 18+, else undici or node-fetch
import fs from 'fs';
import path from 'path';
import dotenv from 'dotenv';
import { fileURLToPath } from 'url';

// Load environment variables
dotenv.config();

const OLLAMA_URL = process.env.OLLAMA_URL || 'http://localhost:11434/api/chat';
const MODEL = process.env.MODEL || 'llama3.2';

const SYSTEM_PROMPT = `You are an expert QA Automation Engineer.
Your task is to generate detailed test cases based on the user's requirement.
Output MUST be a strictly valid JSON object.`;

async function generate(userInput) {
    const messages = [
        { role: 'system', content: SYSTEM_PROMPT },
        { role: 'user', content: userInput }
    ];

    try {
        console.log(`Sending request to ${OLLAMA_URL} with model ${MODEL}...`);
        const response = await fetch(OLLAMA_URL, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                model: MODEL,
                messages: messages,
                stream: false,
                format: 'json'
            })
        });

        if (!response.ok) {
            throw new Error(`Ollama API Error: ${response.status} ${response.statusText}`);
        }

        const data = await response.json();
        return data.message.content;
    } catch (error) {
        console.error("Failed to generate test cases:", error.message);
        process.exit(1);
    }
}

// CLI Execution
const args = process.argv.slice(2);
if (args.length === 0) {
    console.log("Usage: node tools/generate_test_cases.js \"<prompt>\"");
    process.exit(0);
}

const prompt = args[0];
generate(prompt).then(output => {
    console.log("\nGenerated Test Cases:");
    console.log(output);

    // Save to .tmp as per Layer 3 protocol
    const tmpDir = path.join(process.cwd(), '.tmp');
    if (!fs.existsSync(tmpDir)) fs.mkdirSync(tmpDir);

    const filename = path.join(tmpDir, `test_cases_${Date.now()}.json`);
    fs.writeFileSync(filename, output);
    console.log(`\nSaved to ${filename}`);
});
